### DISCUSSION TITLE
docToolchain using .js versions - could that make sense?

### DISCUSSION THREAD

**Discussion by joel-loewenstein** [docToolchain/docToolchain#552] [Ideas]
Hi, while our discussion about asciidoctor-pdf you made me aware of asciidoctor-web-pdf, which was previously called asciidoctor-pdf.js but uses a totally different engine then asciidoctor and to remove confusion to asciidoctor-pdf they renamed it to asciidoctor-web-pdf.

Anyway, it is node.js based. If you would want to use it then you need a node.js running. But if you do, it instantly pops into my mind that i might want to run asciidoctor.js since i have the node anyway.

There is a gradle plugin to include node.js into your build. I have not a full picture yet, just letting my thoughts pop out.

**Comment by joel-loewenstein** [comment]
I just talked to @mogztter in the kroki [community forum](https://gitter.im/kroki-project/community), if there is was .js version of kroki ever discussed, but he pointed out due to the technical components underlying enabling the different diagramming languages this was never condisdered.

So for now i have to burry my idea of maintaining docToolchain on one runtime... but still as you mentioned including asciidoctor-web-pdf might make sense, and kroki anyway. @mogztter pointed out the diagramming plugin supports kroki, which is another topic i have to learn now on my list... soo much to learn still...

**Reply by rdmueller** [reply]
I would love to add kroki to the docker image to make it even easier to use.

**Reply by joel-loewenstein** [reply]
a local java install is documented, so that should be possible i assume

**Reply by ggrossetie** [reply]
Kroki is also available as a "fat-jar" so another option is to start the Kroki server using `java -jar` (as described in: https://docs.kroki.io/kroki/setup/manual-install/#_download_and_run_kroki)

Please note that in this case, "only" the following diagram libraries will be available:
- Bytefield
- Ditaa
- Erd
- GraphViz
- Nomnoml
- PlantUML
- C4 model (with PlantUML)
- Svgbob
- UMlet
- Vega
- Vega-Lite
- WaveDrom

If you want to use BlockDiag, ActDiag, NwDiag, SeqDiag, Mermaid, Excalidraw or BPMN, then you will also need to start companion containers (Docker). Namely:

- `yuzutech/kroki-mermaid`
- `yuzutech/kroki-bpmn`
- `yuzutech/kroki-excalidraw`

**Reply by joel-loewenstein** [reply]
thanks for pointing that out, excalidraw explicitly raised my personal interest...

**Reply by rdmueller** [reply]
oh, the fat-jar would make it easy to integrate. I think I will go down this route first. This would also allow for an easy local install.

**Comment by rdmueller** [comment]
the node part of asciidoctor-web-pdf seems to be  very thin. Basically, afaik it adds some polyfills to the html output, starts chromium headless and does a print-to-pdf. So, we might be able to avoid node.

but we also might use the plugin you where writing about. I guess, doctoolchain would still run without node if the plugin is not used...

**Reply by joel-loewenstein** [reply]
he was acutally referring to the asciidoctor diagramming plugin. but i might have misunderstood that.

**Reply by ggrossetie** [reply]
> the node part of asciidoctor-web-pdf seems to be very thin. Basically, afaik it adds some polyfills to the html output, starts chromium headless and does a print-to-pdf. So, we might be able to avoid node.

That's partially true 😉 
The main point pain will probably to start and control the Chromium headless instance. Asciidoctor Web PDF is using a Node library called Puppeteer (maintained by Google) so it's pretty straightforward.
Starting Chromium headless and opening a page might not be an issue but you will probably need to find a Java/Ruby library to instruct the Chromium instance to print to PDF (probably using the DevTools protocol: https://chromedevtools.github.io/devtools-protocol/).
In Ruby, you can probably use: https://github.com/rubycdp/ferrum
In Java, I'm not sure...

The converter is also doing some logic but that could be re-implemented in another programming language.
We are also using a Node library called https://pdf-lib.js.org/ to add additional metadata and fix some issues after the PDF has been generated by Chromium.

**Reply by rdmueller** [reply]
I knew there was some more magic behind the scenes :-)
The results of asciidoctor-web-pdf are great. If only we could bring the dependencies down a bit...

**Comment by rdmueller** [comment]
so, my goal is at the moment
- to finish work on 2.0.0 and release it. There are only some easy issues left
- then update dependencies like upgrade to ascciidoctor 2.x, asciidoctor-pdf latest version, reveals js etc.
- then add some new features like kroki and asciidoctor-web-pdf
- I guess in these last steps, we will have to first include the node.js dependency

since the architecture for the tasks is modular, I see no reason why we shouldn't already work on the asciidoctor-web-pdf version... let's add it!
